File: hw5/assignment/SplitFileServer.go
/*
SplitFileServer.go
20190532 sang yun lee
*/

package main

import (
	"bufio"
	"fmt"
	"io"
	"net"
	"os"
	"strconv"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Println("This program only accepts one argument")
		os.Exit(1)
	}
	//server Port
	serverPort := os.Args[1]

	// listen from client request
	listener, err := net.Listen("tcp", ":"+serverPort)
	if err != nil {
		fmt.Println("fail to start server: ", err)
		return
	}
	defer func(listener net.Listener) {
		err := listener.Close()
		if err != nil {
			fmt.Println("fail to close listener: ", err)
		}
	}(listener)
	fmt.Println("Waiting for connections...")

	// main loop
	for {
		// accept to client request
		conn, err := listener.Accept()
		if err != nil {
			fmt.Println("fail to accept client request: ", err)
			continue
		}

		// get command from client
		commandBuffer := make([]byte, 1024)
		read, err := conn.Read(commandBuffer)
		if err != nil {
			_, err = conn.Write([]byte("fail to read file"))
			if err != nil {
				fmt.Println("Write failed")
				continue
			}
		} else {
			_, err = conn.Write([]byte("ok"))
			if err != nil {
				fmt.Println("Write failed")
				continue
			}
		}
		commandName := string(commandBuffer[:read])

		// put case
		if commandName == "put" {
			// get file Name from client
			fileNameBuffer := make([]byte, 1024)
			read, err = conn.Read(fileNameBuffer)
			if err != nil {
				_, err = conn.Write([]byte("fail to read file"))
				if err != nil {
					fmt.Println("Write failed")
					continue
				}
			} else {
				_, err = conn.Write([]byte("ok"))
				if err != nil {
					fmt.Println("Write failed")
					continue
				}
			}
			fileName := string(fileNameBuffer[:read])
			fmt.Println("Received file: ", fileName)

			// get file size from client
			fileSizeBuffer := make([]byte, 1024)
			read, err = conn.Read(fileSizeBuffer)
			if err != nil {
				_, err = conn.Write([]byte("fail to read file"))
				if err != nil {
					fmt.Println("Write failed")
					continue
				}
			} else {
				_, err = conn.Write([]byte("ok"))
				if err != nil {
					fmt.Println("Write failed")
					continue
				}
			}
			// translate string to int
			fileSize, err := strconv.ParseInt(string(fileSizeBuffer[:read]), 10, 64)
			if err != nil {
				fmt.Println("fail to transfer file size:", err)
				continue
			}

			// create file
			file, err := os.Create(fileName)
			if err != nil {
				fmt.Println("fail to create file:", err)
				continue
			}

			// save file contents
			var receivedBytes int64
			buffer := make([]byte, 1024)
			for {
				n, err := conn.Read(buffer)
				if err != nil {
					if err != io.EOF {
						fmt.Println("fail to read file:", err)
					}
					break
				}
				receivedBytes += int64(n)

				// write content to file
				if _, err := file.Write(buffer[:n]); err != nil {
					fmt.Println("fail to write file:", err)
					break
				}

				// check all content is done
				if receivedBytes >= fileSize {
					break
				}
			}

			fmt.Printf("%s file store sucessful!\n", fileName)
			err = file.Close()
			if err != nil {
				fmt.Println("file closed error")
				continue
			}
		} else if commandName == "get" { // get case
			// get fileName
			fileNameBuffer := make([]byte, 1024)
			read, _ = conn.Read(fileNameBuffer)
			fileName := string(fileNameBuffer[:read])
			// try to open file
			originalFile, err := os.Open(fileName)
			// error occur
			if err != nil {
				_, err = conn.Write([]byte("fail to open file"))
				if err != nil {
					fmt.Println("Write failed")
					continue
				}
			} else {
				// try to get file Stats
				fileInfo, err := originalFile.Stat()
				if err != nil {
					fmt.Println("fail to get file stat:", err)
					return
				}
				// convert file Size string to Integer
				size := strconv.FormatInt(fileInfo.Size(), 10)
				// send file Size to server
				_, err = conn.Write([]byte(size))
				if err != nil {
					fmt.Println("Write failed")
					continue
				}
			}
			fmt.Println("Request from client to send :" + fileName)

			statusBuffer := make([]byte, 1024)
			read, _ = conn.Read(statusBuffer)
			if string(statusBuffer[:read]) != "ok" {
				continue
			}

			// get Buffer from file
			reader := bufio.NewReader(originalFile)
			for {
				// get 1 byte from file
				b, err := reader.ReadByte()
				if err != nil {
					break // finish to reach file end
				}
				_, err = conn.Write([]byte{b})
				if err != nil {
					fmt.Println("Write Error")
					break
				}
			}
			fmt.Printf("%s send successful\n", fileName)
		} else {
			fmt.Println("unknown command: ", commandName)
		}
	}
}

File: hw5/rust_bonus_assignment/src/SplitFileServer.rs
/*
SplitFileServer.rs
20190532 sang yun lee
*/

use std::env::args;
use std::fs::File;
use std::io::{Read, Write, BufReader};
use std::net::TcpListener;
use std::process::exit;

// message buffer size
const MSG_SIZE: usize = 1024;

fn main() {
	// get system argument
	let args: Vec<String> = args().collect();
	if args.len() != 2 {
		println!("Invalid argument");
		exit(1);
	}
	// serer port from system argument
	let server_port = &args[1];

	// create server_socket
	let server_socket = TcpListener::bind(format!("0.0.0.0:{}", server_port)).expect("Listener failed to bind");
	println!("Waiting for connections...");

	loop {
		// set client connection
		if let Ok((mut socket, _)) = server_socket.accept() {
			// read from client
			let mut command_buffer = vec![0; MSG_SIZE];
			socket.read(&mut command_buffer).unwrap();
			let command_name = command_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
			let command_name = String::from_utf8(command_name).expect("invalid utf8 message");
			// write to client
			socket.write("ok".as_bytes()).unwrap();

			// put case
			if command_name == "put"{
				//get file Name from client
				// read from client
				let mut filename_buffer = vec![0; MSG_SIZE];
				socket.read(&mut filename_buffer).unwrap();
				let file_name = filename_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
				let file_name = String::from_utf8(file_name).expect("invalid utf8 message");
				// write to client
				socket.write("ok".as_bytes()).unwrap();

				// get file size from client
				// read from client
				let mut filesize_buffer = vec![0; MSG_SIZE];
				socket.read(&mut filesize_buffer).unwrap();
				let filesize_string = filesize_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
				let filesize_string = String::from_utf8(filesize_string).expect("invalid utf8 message");
				// write to client
				socket.write("ok".as_bytes()).unwrap();
				let file_size = match filesize_string.parse() {
					Ok(num) => num,
					Err(err) => {
						eprintln!("convert failed : {}", err);
						0
					}
				};

				// file create
				match File::create(&file_name) {
					Ok(mut file) => {
						let mut received_bytes = 0;
						let mut buffer = vec![0; MSG_SIZE];
						while received_bytes < file_size {
							match socket.read(&mut buffer) {
								Ok(n) => {
									if n == 0 {
										break;
									}
									file.write_all(&buffer[..n]).unwrap();
									received_bytes += n;
								},
								Err(e) => {
									println!("Error occurred: {:?}", e);
									continue;
								}
							}
						}
					},
					Err(error) => {
						eprintln!("File open Error: {}", error);
						exit(1);
					}
				};
				println!("{} save successful!", file_name);
			} else if command_name == "get" {
				//get file Name from client
				// read from client
				let mut filename_buffer = vec![0; MSG_SIZE];
				socket.read(&mut filename_buffer).unwrap();
				let file_name = filename_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
				let file_name = String::from_utf8(file_name).expect("invalid utf8 message");

				// try to open file
				let src_file = match File::open(&file_name) {
					Ok(file) => {
						// try to get file metadata
						match file.metadata() {
							Ok(metadata) => {
								// send file size to client
								socket.write(metadata.len().to_string().as_bytes()).unwrap();
								file
							},
							Err(err) => {
								eprintln!("fail to get file size: {}", err);
								socket.write("error!".as_bytes()).unwrap();		
								continue;
							}
						}
					},
					Err(err) => {
						eprintln!("fail to open file: {}", err);
						socket.write("error!".as_bytes()).unwrap();
						continue;
					}
				};

				println!("Request from client to send : {}", &file_name);
				// read from client
				let mut status_buffer = vec![0; MSG_SIZE];
				socket.read(&mut status_buffer).unwrap();
				let status_res = status_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
				let status_res = String::from_utf8(status_res).expect("invalid utf8 message");
				// client status check
				if status_res != "ok" {
					eprintln!("fail to receive file size");
					continue;
				}

				// set Reader Buffer in file
				let mut reader = BufReader::new(src_file);
				let mut buffer = [0; 1];
				loop {
					// read 1 byte from file
					match reader.read(&mut buffer) {
						// EOF
						Ok(0) => {
							println!("{} send successful!", file_name);
							break
						},
						Ok(_) => {
							let byte = buffer[0];
							socket.write(&[byte]).unwrap();
						}
						Err(error) => {
							eprintln!("read failed: {}", error);
							break;
						}
					}
				}
			} else {
				println!("Unknown command!: {}", command_name);
			}
		}
	}
}